### ISSUE TITLE
userdel -r -f ... can remove random files under '/'

### ISSUE THREAD

**Issue by wolfsage** [shadow-maint/shadow#1050]
Howdy.

It appears that userdel has no protection against erasing a system if a user's home dir is set to '/'. (A lot of users like systemd-network, etc have their homedir set to '/').

If you remove one of them with userdel -r -f, one of two things might happen:

* userdel may open /proc first, attempt to delete something in there, fail, and abort, and you are safe
* userdel may open some other directory first (like /usr!), delete everything in there, and move on, trashing your system.

I can't imagine a good reason userdel should ever attempt to delete / and all files under it. Can this be prevented somehow?

I've seen this issue on Debian 6.1.94-1, and it seems highly repeatable (using a DigitalOcean Debian 12 Droplet).

For some reason, on Debian 6.1.38-2 and earlier, userdel always seems to find /proc first, and never deletes '/'.

This is Debian shadow-utils 4.13 in both cases

**Comment by alejandro-colomar** [comment]
Hmmm, some question that may be dumb:

Does systemd-network (or any user) need to have its homedir set to `/`?

**Comment by alejandro-colomar** [comment]
> Debian 6.1.94-1

What is that?  Debian 6 was dead many years ago.

**Comment by wolfsage** [comment]
Woof, sorry, that's likely the kernel version
```
$ uname -a
Linux 6.1.0-22-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.94-1 (2024-06-21) x86_64 GNU/Linux
```

This is debian bookworm
```
$ cat /etc/debian_version 
12.6
```

**Comment by alejandro-colomar** [comment]
Why should a user have its homedir set to `/`?

**Comment by wolfsage** [comment]
It certainly seems weird to me, but adding this protection still seems like a good thing to do.

**Comment by alejandro-colomar** [comment]
> It certainly seems weird to me,

Please report a bug to systemd (or maybe it's a distro issue?), and please add a link here to see what they say.

> but adding this protection still seems like a good thing to do.

I'm not sure.  Those are heuristics which increase the complexity of the software, possibly adding bugs, and I'm not sure we can write correctly those heuristics.  Should we also try to prevent removing /usr/bin? or /usr/sbin?  Where do we stop?

I'd rather keep it raw, and let one run `rm -rf /`.  If they broke their systems, so that userdel runs `rm -rf /`, it's not our business.  userdel is a symptom, not the cause.

**Comment by wolfsage** [comment]
> I'm not sure. Those are heuristics which increase the complexity of the software, possibly adding bugs, and I'm not sure we can write correctly those heuristics.

I agree that more code means more possibility for bugs, but are you really suggesting that "Does path X resolve to '/'?" even in the simplest form is too hard to do? Or am I missing something?

Even a check of "does the string user_home contain the bytes "/\0" and nothing more" would be a great start.


> Should we also try to prevent removing /usr/bin if it's homedir? or /usr/sbin? Where do we stop?

I'm not sure why you're jumping to a slippery slope argument. Surely no one wants to ever rm -rf / based on some other tool trying to clean up something. Beyond that, no, I wouldn't say more protections are needed.

> I'd rather keep it raw, and let one run rm -rf /. If they broke their systems, so that userdel runs rm -rf /, it's not our business. userdel is a symptom, not the cause.

I think most users would be much happier if their systems didn't mysteriously break because the tools they use come with no guardrails, and the more tools that are improved to be less dangerous the happier we will all be. I do not understand the resistance to not destroying things for users based on a mistake.

Anyway, please close this if you do not agree, thanks.

**Comment by wolfsage** [comment]
> Please report a bug to systemd (or maybe it's a distro issue?), and please add a link here to see what they say.

https://github.com/systemd/systemd/issues/33743

Cheers.

**Comment by poettering** [comment]
I am pretty sure userdel should check if homedir is actually owned by the UID/GID of the account. If not, it should stay away from the homedir (and maybe print a warning message saying so)

**Comment by hallyn** [comment]
> Anyway, please close this if you do not agree, thanks.

No, don't close this :)

I'm in agreement with Lennart.  My only concern there is, there are packages that create 'homedirs' for system users which are then owned by root or another user.  If we suddenly stop removing such dirs, it may break some things (like package re-installs).  Printing a warning doesn't help as a lot of this will be hidden behind scripts and automation.

**Comment by alejandro-colomar** [comment]
While checking <https://github.com/shadow-maint/shadow/issues/1062>, I've learnt that we actually have some tests to check that would prevent removing "/", or someone else's home.  Those tests acknowledge that this is rather imperfect, as expected but the tests are in place.

However, using `-f` skips some of those tests.  Also, some tests are conditionally compiled out (a comment in the source says it's because they're slow.  The distro is responsible for compiling them, I guess.

**Comment by hallyn** [comment]
> While checking #1062, I've learnt that we actually have some tests to check that would prevent removing "/", or someone else's home. Those tests acknowledge that this is rather imperfect, as expected but the tests are in place.
> 
> However, using `-f` skips those tests. Also, some tests are conditionally compiled out (a comment in the source says it's because they're slow. The distro is responsible for compiling them, I guess.

Ok, but '-f' should not skip those tests.  -f in userdel is meant to be about whether or not to do the deletion of the user.  But if HOME is /, userdel -f should not try to remove it.  The !fflag check there should be removed.  Do you mind posting a minimal patch for that?

**Comment by alejandro-colomar** [comment]
> Ok, but '-f' should not skip those tests. -f in userdel is meant to be about whether or not to do the deletion of the user. But if HOME is /, userdel -f should not try to remove it. The !fflag check there should be removed.

According to the documentation (manual page), it should skip those tests:

```
     -f, --force
         This option forces the removal of the user account,
         [...]. It also forces
         userdel to remove the user's home directory and mail
         spool, even if another user uses the same home
         directory [...]

         Note: This option is dangerous and may leave your
         system in an inconsistent state.

```

Or do you mean we should also change that?

> Do you mind posting a minimal patch for that?

**Comment by alejandro-colomar** [comment]
Maybe specifying the option once for "remove the user if still loged in" and similar danger levels and twice for "use nuclear bombs if necessary"?

Or maybe we want more levels?

How about this?:

```diff
diff --git i/src/userdel.c w/src/userdel.c
index ead69604..53a7c600 100644
--- i/src/userdel.c
+++ w/src/userdel.c
@@ -79,7 +79,7 @@ static uid_t user_id;
 static gid_t user_gid;
 static char *user_home;
 
-static bool fflg = false;
+static int fflg = 0;
 static bool rflg = false;
 #ifdef WITH_SELINUX
 static bool Zflg = false;
@@ -312,7 +312,7 @@ static void remove_usergroup (void)
                return;
        }
 
-       if (!fflg) {
+       if (fflg < 1) {
                /*
                 * Scan the passwd file to check if this group is still
                 * used as a primary group.
@@ -832,7 +832,7 @@ static int remove_mailbox (void)
                }
        }
 
-       if (fflg) {
+       if (fflg >= 1) {
                if (unlink (mailfile) != 0) {
                        fprintf (stderr,
                                 _("%s: warning: can't remove %s: %s\n"),
@@ -998,7 +998,7 @@ int main (int argc, char **argv)
                                         long_options, NULL)) != -1) {
                        switch (c) {
                        case 'f':       /* force remove even if not owned by user */
-                               fflg = true;
+                               fflg++;
                                break;
                        case 'h':
                                usage (E_SUCCESS);
@@ -1131,7 +1131,7 @@ int main (int argc, char **argv)
         * a cron job may be started on her behalf, etc.
         */
        if ((prefix[0] == '\0') && !Rflg && user_busy (user_name, user_id) != 0) {
-               if (!fflg) {
+               if (fflg < 1) {
 #ifdef WITH_AUDIT
                        audit_logger (AUDIT_DEL_USER, Prog,
                                      "deleting user logged in",
@@ -1160,7 +1160,7 @@ int main (int argc, char **argv)
                                 _("%s: %s home directory (%s) not found\n"),
                                 Prog, user_name, user_home);
                        rflg = 0;
-               } else if ((0 == home_owned) && !fflg) {
+               } else if ((0 == home_owned) && fflg < 2) {
                        fprintf (stderr,
                                 _("%s: %s not owned by %s, not removing\n"),
                                 Prog, user_home, user_name);
@@ -1172,7 +1172,7 @@ int main (int argc, char **argv)
 
 #ifdef EXTRA_CHECK_HOME_DIR
        /* This may be slow, the above should be good enough. */
-       if (rflg && !fflg) {
+       if (rflg && fflg < 2) {
                struct passwd *pwd;
                /*
                 * For safety, refuse to remove the home directory if it
```

**Comment by alejandro-colomar** [comment]
Oops, I one of the changes above is better moved to level 2:

```diff
diff --git i/src/userdel.c w/src/userdel.c
index 53a7c600..4ee51acb 100644
--- i/src/userdel.c
+++ w/src/userdel.c
@@ -832,7 +832,7 @@ static int remove_mailbox (void)
                }
        }
 
-       if (fflg >= 1) {
+       if (fflg >= 2) {
                if (unlink (mailfile) != 0) {
                        fprintf (stderr,
                                 _("%s: warning: can't remove %s: %s\n"),
```

**Comment by hallyn** [comment]
yeah, '-f -f' would suffice imo.

**Comment by alejandro-colomar** [comment]
Ping @wolfsage , @poettering 

We closed the PR because you didn't test it.  Are you still interested in this?  Could you please test it?

**Comment by alejandro-colomar** [comment]
@bluca, maybe?
